### PR TITLE
Axel's weather app submission

### PR DESCRIPTION
**Pull request by AxelHammarback**
**Link to my deployed project:**
http://axel-weather.netlify.app/

**My reflections on how this week's project turned out:**
I'm very happy with my work and how the way this week's project turned out. It a struggle initially, but when I understood the fetch concepts and how to work with the data, it became incredibly fun. Also, CSS is hard.

**Things I'd like to have clarified or explained in more detail:**
I still struggle with looping through the contents of an object. Is that possible?

### CODE REVIEW

File: code/style.css [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  /* Make the text look smoother: */
+  text-rendering: optimizeLegibility;
+  -webkit-font-smoothing: antialiased;
+  -moz-osx-font-smoothing: grayscale;
+
+  margin: 0;
+  background-color: #0C60F2;
+  color: rgba(255,255,255,0.87);
+  font-family: 'Work Sans', sans-serif;
+  display: flex;
```
Comment (VanessaSue27) [inline code/style.css:27]: The body is set to `display: flex`
This doesn't seem to have any effect when I tested it on dev tools.
Maybe this line could be removed?

File: code/style.css [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+
+/* THIS IS THE GUY THAT HOLDS EVERYTHING IN PLACE */
+.container {
+  width: 100%;
+  margin: auto;
```
Comment (VanessaSue27) [inline code/style.css:34]: This container is set to have a width of 100%, so `margin: auto` doesn't seem to have any effect
Maybe it can be ommitted?

File: code/style.css [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+  display: flex;
+  flex-direction: column;
+  align-content: space-between;
+  position: relative;
```
Comment (VanessaSue27) [inline code/style.css:38]: Super good that you use position in your project - I shy away from using most of the time 🙈

File: code/style.css [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+  overflow: hidden;
+  position: fixed;
```
Comment (VanessaSue27) [inline code/style.css:40]: This one container has two position rules: one set to relative and one set to fixed (It seems fixed is the one being ignored, so maybe it can be removed?)

File: code/style.css [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+  z-index: 1;
+  display: none;
+  align-items: center;
+}
+
+.header-content {
+  max-width: 375px;
+  margin: 0 auto;
+  display: flex;
+  align-items: center;
```
Comment (VanessaSue27) [inline code/style.css:73]: This one is being set to `display: flex` but no direction is specified, so `align-items: center;` doesn't seem to be having any effect, maybe it's not necessary to have that line?

File: code/style.css [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+  box-shadow: 0px 3px 0px #0C60F2;
+  border: 2px solid #0C60F2;
+}
+
+.button-primary:hover {
+  background-color: white;
+  border: 4px solid #0C60F2;
+}
+
+.button-primary:focus {
```
Comment (VanessaSue27) [inline code/style.css:120]: Unsure about the difference between hover and focus in this case (I will need some extra insight help here)
But they both have the same properties: 
```
background-color: white;
 border: 4px solid #0C60F2;
```
Can we avoid having to declare them twice? (same for lines 135-140)

File: code/style.css [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+.button-secondary:active {
+  background-color: #ffffff22;
+}
+
+button {
+  margin-left: 8px;
+  padding: 0px;
+  width: 36px;
+  height: 34px;
+  box-sizing: border-box;
```
Comment (VanessaSue27) [inline code/style.css:154]: Okay, one of the only times I see border-box being used effectively. I was not really clear why I would use border-box, but I think seeing your buttons and the effect you have created using this property, I understand what can be achieved with it 🤩 NICE!

File: code/style.css [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+  display: flex;
+  flex-direction: column;
+  justify-content: space-between;
+  flex-grow: 1;
+  max-width: 375px;
+  margin: 0 auto;
+}
+
+.image-hero-weather-container {
+  animation: MoveUpDown 4s ease-in-out infinite;
```
Comment (VanessaSue27) [inline code/style.css:194]: Animation, yessss!!! 💥

File: code/style.css [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+.image-hero-weather-container img {
+  width: 250px;
+  height: 250px;
+}
+
+.main-weather-top {
+  /* border: 1px dashed salmon; */
+  flex-direction: row;
```
Comment (VanessaSue27) [inline code/style.css:216]: Here this `flex-direction` doesn't seem to be having an effect, can it be avoided?

File: code/style.css [hunk 10]
```diff
@@ -0,0 +1,378 @@
+/* I left the "border: 1px dashed" lines due to debugging purposes. */
```
Comment (VanessaSue27) [inline code/style.css:1]: Really good tip! I also use these borders when styling to know what is going on with each element 👌

File: code/style.css [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+  line-height: 150px;
+}
+
+
+/* (This is the celsius sign) */
+.temp-big-but-slightly-smaller {
+  vertical-align: super;
+  font-size: 42px;
+  font-weight: 300;
+  vertical-align: top;
```
Comment (VanessaSue27) [inline code/style.css:268]: `vertical-align` being declared twice here, when testing it I found `vertical-align: super` is the one being ommitted, so maybe that one can be removed

File: code/style.css [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+
+/* 
+Breakpoints:
+<600      SMARTPHONE
+600-1024  TABLET
+>1024     DESKTOP
+*/
+
+
+/* "The one key thing to remember when trying to position a child div relative to it’s parent is that the child should be given the CSS property position:absolute; and the parent set to either position:absolute; or position:relative;" */
```
Comment (VanessaSue27) [inline code/style.css:345]: AMAZING reminder! 🎉

File: code/style.css [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+main {
+  max-height: 580px;
+}
+
+footer {
+  flex-grow: 1;
+}
+
+.image-hero-weather-container {
+  position: absolute;  right: -20px;
```
Comment (VanessaSue27) [inline code/style.css:364]: `right: -20px`
can start in the next line instead

File: code/style.css [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+
+main {
+  max-height: 580px;
+}
+
+footer {
+  flex-grow: 1;
+}
+
+.image-hero-weather-container {
```
Comment (VanessaSue27) [inline code/style.css:363]: The styling for `image-hero-weather-container` is already declared on line 193.
I couldn't find anything particular, but is there a reason why it needs to be delcared here again? Nothing in the styling seems to have changed from the original declaration~

**Review by VanessaSue27 — APPROVED** [review]
Hello hello!
I learned so much by reviewing your code, Axel! 😎

I can tell you do a lot of research and try to implement new methods, nice to see that interest to find new tools!
I have also noticed the eye for detail and design here: there are so many small effects carefully chosen and implemented, I'm very impressed by that. Especially the header section with the hover effects on buttons.
I admire that you implemented this City Selector input, it could not have been so simple and the result is great!
As I mentioned above, all the color changes, hover effects, and ALSO the animation on the weather icon, makes this weather app stand out. I loved discovering all those small details 🤩

I also had a good laugh reading through the commit messages 😝 It was like reading a funny journal, and I can agree: CSS - the struggle is real! 💪

I noticed the .gitignore file missing - nice to add it so you avoid the DS_Store files.

I will need some help from the coach later on regarding the SVG jQuery file: I heard from some team mates that they were struggling to get the icons to work because they were in SVG format - so I'm assuming this was your way to work around it? But since I have no knowledge on jQuery, not sure which feedback to give here~~

All in all a really interesting project to review, and I am glad you took the time to understand and make fetch() work 💪 It was a process for all of us, but now we can see the great results! 💥

I have left a lot of comments, but I don't find anything crucial that needs to be changed (regarding indentation and variable namig and so on), so I'm going to mark it as Approved and wait for the coach to get back to us 😁

My pleasure reviewing your code,
Vanessa ~~

File: code/index.html [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    </div>
+  </footer>
+
+
+
+
+</div>
+  <script src="script.js" defer></script>
+  <script src="countryTransform.js" defer></script>
+  <script src="svgColor.js" defer></script>
```
[the remark this answers is not in the document]
Comment (AxelHammarback) [reply code/index.html:95]: Ah yes, need to remember to add the dot. Thanks, Vanessa!

And in this case, it's (probably) not needed since they're in the bottom of the HTML, but I just try to make it a habit to add "defer". (If a code coach reading this can think of a reason not to, feel free to comment here.)

File: code/script.js [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+// Function takes a Kelvin value, and returns a celsius value. 
+const kelvinToCelsius = (temp) => {
+  let celsius = Math.round(temp-273.15);
+  return celsius
+}
+
+// Returns the big weather image depending on the weather.
+const fetchWeatherImage = (weather) => {
```
[the remark this answers is not in the document]
Comment (AxelHammarback) [reply code/script.js:51]: initially they were added for debugging purposes – but later on, I thought "Hmm, maybe I should print a nicer-looking message to the DOM". I ended up not printing them in the DOM, but left them if I want to revisit this some day

File: code/style.css (hunk 11 above)
Comment (AxelHammarback) [reply code/style.css:268]: Awesome, good find – thanks!

File: code/style.css (hunk 12 above)
Comment (AxelHammarback) [reply code/style.css:345]: Ah, yes – found it here: https://tomelliott.com/html-css/css-position-child-div-parent

But left it as a comment to try to remember it better :D

File: code/style.css (hunk 14 above)
Comment (AxelHammarback) [reply code/style.css:363]: Nope, I just missed it due to sloppiness 😅

File: README.md [hunk 17]
```diff
@@ -1,13 +1,23 @@
...
+An app which displays the current weather of any city in the world, as well as a five-day forecast.
 
 ## The problem
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+The first point of action was to successfully fetch the data from OpenWeatherMap. I looked at their  API to see what information which could be extracted, and selected a few points of data which I would want to use in my app. 
 
-## View it live
+After I figured out how to do populate the DOM with data, I started with the styling part. For this, I used Figma, where I also documented my process. See it [here](https://www.figma.com/file/GpbiZuim0lWaqhj3KGM3ZZ/Weather-application?node-id=0%3A1). 
+
```
[the remark this answers is not in the document]
Comment (EmOlsz) [reply README.md:14]: Indeed :)

File: code/countryTransform.js [hunk 18]
```diff
@@ -0,0 +1,258 @@
+// Found a ready-made function which takes a country code and spits out the country name.
+// https://gist.github.com/maephisto/9228207
+
+var isoCountries = {
```
Comment (EmOlsz) [inline code/countryTransform.js:4]: Very ambitious Axel, well done solid job!

File: code/index.html [hunk 19]
```diff
@@ -0,0 +1,97 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <title>Weather app</title>
+  <link rel="stylesheet" href="style.css">
```
Comment (EmOlsz) [inline code/index.html:7]: `./`

File: code/index.html [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+  <link href="https://fonts.googleapis.com/css2?family=Work+Sans:wght@100;200;300;400;600;800&display=swap" rel="stylesheet">
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
+  
+</head>
+
+
+<body>
+
+
```
Comment (EmOlsz) [inline code/index.html:16]: A lot of empty lines, let's delete them. For the future - one empty line between elements in considered maximum

File: code/index.html (hunk 15 above)
[the remark this answers is not in the document]
Comment (EmOlsz) [reply code/index.html:95]: Well, since you add your scripts at the end of HTML, there is no need to add defer from logical point of view, but it's no harm that you added them anyway ;)

File: code/index_backup.html [hunk 21]
```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1.0">
+  <title>Weather app</title>
+  <link rel="stylesheet" href="style.css">
```
Comment (EmOlsz) [inline code/index_backup.html:7]: `./`

File: code/index_backup.html [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+      <img src="./assets/ic_cloudy.svg" id="weatherIcon" class="svg weather-icon" alt="weather icon"/>
+
+      <h2>5-day forecast</h2>
+      <section id="forecastBox"></section>
+      
+  </main>
+  <script src="script.js" defer></script>
+  <script src="countryTransform.js" defer></script>
+  <!-- <script src="svgColor.js" defer></script> -->
```
Comment (EmOlsz) [inline code/index_backup.html:49]: Before final push, always remember to delete all commented code

File: code/response.js [hunk 23]
```diff
@@ -0,0 +1,92 @@
+// Fields in API response
```
Comment (EmOlsz) [inline code/response.js:1]: Hmm, I don't think you need this file ;)

File: code/script.js [hunk 24]
```diff
@@ -0,0 +1,204 @@
...
+const tempFeelsLike = document.getElementById("tempfeelslike");
+const description = document.getElementById("description");
+const sunrise = document.getElementById("sunrise");
+const sunset = document.getElementById("sunset");
+const weatherType = document.getElementById("weathertype");
+
+// This one toggles whether to show the city selector or not, by adding a display: none;
+const toggleCountrySelector = () => {
+  var citySelector = document.getElementById("citySelector");
+  var inputCity = document.getElementById("inputCity")
```
[the remark this answers is not in the document]
Comment (EmOlsz) [reply code/script.js:19]: Yes, definitely!

File: code/script.js [hunk 25]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+// Returns the big weather image depending on the weather.
+const fetchWeatherImage = (weather) => {
+  // Purpose of this function should be to fetch an image and load it to the DOM, depending on the weather type.
+  // Available: Clouds, Clear, Snow, Rain, Drizzle, Thunderstorm. (MIST IS MISSING)
+  const image = document.getElementById("weatherImage");
+
+  if(weather === 'Clouds') {
+    console.log("It's cloudy today.");
```
Comment (EmOlsz) [inline code/script.js:57]: A lot of console.logs - we should delete all of them before final push

File: code/style.css (hunk 2 above)
Comment (EmOlsz) [reply code/style.css:34]: Depends if this `margin: auto` is used later on in different resolutions

File: code/style.css [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+  padding: 16px;
+}
+
+.header-city-selector {
+  background-color: #0C60F2;
+  padding: 20px 16px 16px 16px;
+  position: absolute;
+  right: 0;
+  left: 0;
+  top: 0;
```
Comment (EmOlsz) [inline code/style.css:62]: Proper order - top, right, bottom, left

File: code/style.css [hunk 27]
```diff
@@ -0,0 +1,378 @@
...
+input:active {
+  outline: 0;
+}
+
+input:focus {
+  outline: 0;
+}
+
+
+
```
Comment (EmOlsz) [inline code/style.css:108]: A lot of empty lines

File: code/style.css [hunk 28]
```diff
@@ -0,0 +1,378 @@
...
+  border: 4px solid white;
+}
+
+.button-secondary:active {
+  background-color: #ffffff22;
+}
+
+button {
+  margin-left: 8px;
+  padding: 0px;
```
Comment (EmOlsz) [inline code/style.css:151]: `padding: 0;`

File: code/style.css [hunk 29]
```diff
@@ -0,0 +1,378 @@
...
+  /* border: 1px dashed goldenrod; */
+  height: 212px;
+  display: flex;
+  flex-direction: column;
+  position: relative;
+}
+
+.footer-top {
+  height: 24px;
+  background-image: url(./assets/clouds-top-border.svg);
```
Comment (EmOlsz) [inline code/style.css:283]: It's safer to wrap url in quotes, some browsers might have problem with reading path without them :

```
background-image: url("./assets/clouds-top-border.svg");
```

File: code/svgColor.js [hunk 30]
```diff
@@ -0,0 +1,35 @@
+// Source: https://stackoverflow.com/questions/11978995/how-to-change-color-of-svg-image-using-css-jquery-svg-image-replacement
```
Comment (EmOlsz) [inline code/svgColor.js:1]: Good job with outer library!

**Review by EmOlsz — CHANGES_REQUESTED** [review]
Well done, solid project Axel! I like your solutions to our requirements, you've done a lot of extra stuff :)
There are some things I would like you to fix, please, take a look on my comments and provide changes.
Apart from that, really nice job!

**Comment by AxelHammarback** [comment]
Hey @EmOlsz! Sorry for the very late fix. Really good comments, I agree with them all and changed the app according to the review. Take care, see you tomorrow 👋

**Review by EmOlsz — APPROVED** [review]
